File: identhub/src/main/java/de/solarisbank/identhub/contract/sign/ContractSigningFragment.java
package de.solarisbank.identhub.contract.sign;

import android.graphics.drawable.LevelListDrawable;
import android.os.Bundle;
import android.text.TextWatcher;
import android.view.FocusFinder;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import com.jakewharton.rxbinding2.widget.RxTextView;

import java.util.ArrayList;
import java.util.List;

import de.solarisbank.identhub.ViewModelFactory;
import de.solarisbank.identhub.R;
import de.solarisbank.identhub.databinding.FragmentContractSigningBinding;
import de.solarisbank.identhub.di.LibraryComponent;
import de.solarisbank.identhub.BaseFragment;
import de.solarisbank.identhub.identity.IdentityActivityViewModel;
import de.solarisbank.identhub.ui.DefaultTextWatcher;
import de.solarisbank.identhub.verfication.phone.CountDownTimeState;
import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.disposables.Disposables;
import timber.log.Timber;

import static de.solarisbank.identhub.verfication.phone.VerificationPhoneViewModel.DEFAULT_STATE;
import static de.solarisbank.identhub.verfication.phone.VerificationPhoneViewModel.DISABLED_STATE;
import static de.solarisbank.identhub.verfication.phone.VerificationPhoneViewModel.ERROR_STATE;
import static de.solarisbank.identhub.verfication.phone.VerificationPhoneViewModel.SUCCESS_STATE;

public class ContractSigningFragment extends BaseFragment {
    private final List<LevelListDrawable> listLevelDrawables = new ArrayList<>();
    protected ViewModelFactory viewModelFactory;
    private Disposable disposable = Disposables.disposed();
    private FragmentContractSigningBinding binding;
    private ContractSigningViewModel viewModel;
    private IdentityActivityViewModel sharedViewModel;
    private View currentFocusedEditText;

    public static Fragment newInstance() {
        return new ContractSigningFragment();
    }

    @Override
    protected void inject(LibraryComponent component) {
        component.inject(this);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentContractSigningBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initViewModel();
        initViews();
        observeInputsState();
        observableInputs();
        observeTimer();
    }

    private void initViewModel() {
        viewModel = new ViewModelProvider(this, viewModelFactory)
                .get(ContractSigningViewModel.class);
        viewModel.startTimer();

        sharedViewModel = new ViewModelProvider(requireActivity(), viewModelFactory)
                .get(IdentityActivityViewModel.class);
    }

    private void initViews() {
        initFocusListener();
        initTextWatcher();
        initStateOfDigitInputField();

        binding.description.setText(String.format(getString(R.string.verification_phone_description), ""));

        binding.submitButton.setOnClickListener(v -> {
            viewModel.onSubmitButtonClicked();
        });

        binding.sendNewCode.setOnClickListener(v -> {
            viewModel.onSendNewCodeClicked();
        });
    }

    private void initFocusListener() {
        View.OnFocusChangeListener onFocusChangeListener = (view, hasFocus) -> {
            if (hasFocus) {
                currentFocusedEditText = view;
            }
        };

        binding.firstDigit.setOnFocusChangeListener(onFocusChangeListener);
        binding.secondDigit.setOnFocusChangeListener(onFocusChangeListener);
        binding.thirdDigit.setOnFocusChangeListener(onFocusChangeListener);
        binding.fourthDigit.setOnFocusChangeListener(onFocusChangeListener);
        binding.fifthDigit.setOnFocusChangeListener(onFocusChangeListener);
        binding.sixthDigit.setOnFocusChangeListener(onFocusChangeListener);

        binding.firstDigit.requestFocus();
    }

    private void observableInputs() {
        disposable = Observable.combineLatest(
                RxTextView.textChanges(binding.firstDigit),
                RxTextView.textChanges(binding.secondDigit),
                RxTextView.textChanges(binding.thirdDigit),
                RxTextView.textChanges(binding.fourthDigit),
                RxTextView.textChanges(binding.fifthDigit),
                RxTextView.textChanges(binding.sixthDigit),
                this::consumeDigitInputs
        )
                .subscribe(code -> viewModel.onVerificationCodeReady(code), throwable -> {
                    Timber.d("Something went wrong");
                });
    }

    private String consumeDigitInputs(
            CharSequence firstDigit,
            CharSequence secondDigit,
            CharSequence thirdDigit,
            CharSequence fourthDigit,
            CharSequence fifthDigit,
            CharSequence sixthDigit) {

        return String.valueOf(firstDigit) +
                secondDigit +
                thirdDigit +
                fourthDigit +
                fifthDigit +
                sixthDigit;
    }

    private void initTextWatcher() {
        TextWatcher textWatcher = new DefaultTextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                if (after > count && after == 1) {
                    changeFocusToNextInput();
                }
            }
        };

        binding.firstDigit.addTextChangedListener(textWatcher);
        binding.secondDigit.addTextChangedListener(textWatcher);
        binding.thirdDigit.addTextChangedListener(textWatcher);
        binding.fourthDigit.addTextChangedListener(textWatcher);
        binding.fifthDigit.addTextChangedListener(textWatcher);
    }

    private void changeFocusToNextInput() {
        View view = FocusFinder.getInstance().findNextFocus(binding.getRoot(), currentFocusedEditText, View.FOCUS_RIGHT);
        view.requestFocus();
    }

    private void observeInputsState() {
        viewModel.getState().observe(getViewLifecycleOwner(), this::onStateOfDigitInputChanged);
    }

    private void observeTimer() {
        viewModel.getTimer().observe(getViewLifecycleOwner(), this::onDataTimerChanged);
    }

    private void onDataTimerChanged(CountDownTimeState countDownTimeState) {
        if (countDownTimeState != null) {
            binding.newCodeCounter.setVisibility(countDownTimeState.isFinish() ? View.GONE : View.VISIBLE);
            binding.sendNewCode.setVisibility(countDownTimeState.isFinish() ? View.VISIBLE : View.GONE);
            binding.newCodeCounter.setText(String.format(getString(R.string.verification_phone_request_code), countDownTimeState.getCurrentValue()));
        }
    }

    private void initStateOfDigitInputField() {
        listLevelDrawables.add((LevelListDrawable) binding.firstDigit.getBackground());
        listLevelDrawables.add((LevelListDrawable) binding.secondDigit.getBackground());
        listLevelDrawables.add((LevelListDrawable) binding.thirdDigit.getBackground());
        listLevelDrawables.add((LevelListDrawable) binding.fourthDigit.getBackground());
        listLevelDrawables.add((LevelListDrawable) binding.fifthDigit.getBackground());
        listLevelDrawables.add((LevelListDrawable) binding.sixthDigit.getBackground());
    }

    private void onStateOfDigitInputChanged(int state) {
        binding.newCodeCounter.setVisibility(state == DEFAULT_STATE ? View.VISIBLE : View.GONE);
        binding.errorMessage.setVisibility(state == ERROR_STATE ? View.VISIBLE : View.GONE);
        binding.sendNewCode.setVisibility(state == ERROR_STATE || state == DISABLED_STATE ? View.VISIBLE : View.GONE);

        binding.sendNewCode.setEnabled(state != DISABLED_STATE);
        binding.submitButton.setEnabled(state != DISABLED_STATE);
        binding.submitButton.setText(state == DISABLED_STATE ? R.string.verification_phone_status_verifying : R.string.contract_signing_preview_sign_action);

        binding.firstDigit.setEnabled(state != DISABLED_STATE);
        binding.secondDigit.setEnabled(state != DISABLED_STATE);
        binding.thirdDigit.setEnabled(state != DISABLED_STATE);
        binding.fourthDigit.setEnabled(state != DISABLED_STATE);
        binding.fifthDigit.setEnabled(state != DISABLED_STATE);
        binding.sixthDigit.setEnabled(state != DISABLED_STATE);

        for (LevelListDrawable listLevelDrawable : listLevelDrawables) {
            listLevelDrawable.setLevel(state);
        }
        if (state == DEFAULT_STATE) {
            clearInputs();
        }

        if (state == SUCCESS_STATE) {
            sharedViewModel.navigateTo(R.id.action_contractSigningFragment_to_applicationProcessingFragment);
        }
    }

    private void clearInputs() {
        binding.firstDigit.setText(null);
        binding.secondDigit.setText(null);
        binding.thirdDigit.setText(null);
        binding.fourthDigit.setText(null);
        binding.fifthDigit.setText(null);
        binding.sixthDigit.setText(null);
    }

    @Override
    public void onDestroyView() {
        disposable.dispose();
        super.onDestroyView();
    }
}

File: identhub/src/main/java/de/solarisbank/identhub/contract/sign/ContractSigningFragment.kt
package de.solarisbank.identhub.contract.sign

import android.graphics.drawable.LevelListDrawable
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnFocusChangeListener
import android.view.ViewGroup
import android.widget.*
import androidx.core.view.isVisible
import androidx.lifecycle.Observer
import com.jakewharton.rxbinding2.view.RxView
import de.solarisbank.identhub.R
import de.solarisbank.identhub.base.IdentHubFragment
import de.solarisbank.identhub.contract.ContractViewModel
import de.solarisbank.identhub.di.FragmentComponent
import de.solarisbank.identhub.domain.data.dto.ContractSigningState
import de.solarisbank.identhub.verfication.phone.PhoneVerificationViewModel
import de.solarisbank.sdk.core.activityViewModels
import de.solarisbank.sdk.core.viewModels
import de.solarisbank.sdk.core_ui.feature.view.hideKeyboard
import de.solarisbank.sdk.data.dto.MobileNumberDto
import de.solarisbank.sdk.data.entity.CountDownTime
import de.solarisbank.sdk.data.entity.format
import de.solarisbank.sdk.domain.model.result.Event
import de.solarisbank.sdk.domain.model.result.Result
import de.solarisbank.sdk.domain.model.result.data
import de.solarisbank.sdk.domain.model.result.succeeded
import de.solarisbank.sdk.feature.customization.ButtonStyle
import de.solarisbank.sdk.feature.customization.customize
import de.solarisbank.sdk.feature.view.BulletListLayout
import io.reactivex.Observable
import io.reactivex.disposables.Disposables
import timber.log.Timber
import java.util.*

class ContractSigningFragment : IdentHubFragment() {
    private val digitsEditTexts: MutableList<EditText> = ArrayList()
    private var disposable = Disposables.disposed()
    private val listLevelDrawables: MutableList<LevelListDrawable> = ArrayList()
    private val sharedViewModel: ContractViewModel by lazy { activityViewModels() }
    private val viewModel: ContractSigningViewModel by lazy { viewModels() }

    private var codeInput: EditText? = null
    private var currentFocusedEditText: View? = null
    private var description: TextView? = null
    private var sendNewCode: Button? = null
    private var newCodeCounter: TextView? = null
    private var transactionDescription: TextView? = null
    private var errorMessage: TextView? = null
    private var submitButton: Button? = null
    private var progress: ProgressBar? = null
    private var codeProgress: ProgressBar? = null
    private var imageView: ImageView? = null
    private var bulletList: BulletListLayout? = null

    override fun inject(component: FragmentComponent) {
        component.inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.identhub_fragment_contract_signing, container, false)
                .also {
                    codeInput = it.findViewById(R.id.codeInput)
                    description = it.findViewById(R.id.description)
                    sendNewCode = it.findViewById(R.id.sendNewCode)
                    newCodeCounter = it.findViewById(R.id.newCodeCounter)
                    transactionDescription = it.findViewById(R.id.transactionDescription)
                    errorMessage = it.findViewById(R.id.errorMessage)
                    submitButton = it.findViewById(R.id.submitButton)
                    progress = it.findViewById(R.id.progress)
                    codeProgress = it.findViewById(R.id.codeProgress)
                    imageView = it.findViewById(R.id.scratch)
                    bulletList = it.findViewById(R.id.noticeBulletList)
                    setUpBulletList()
                    customizeUI()
                }
    }

    private fun customizeUI() {
        submitButton?.customize(customization)
        sendNewCode?.customize(customization, ButtonStyle.SecondaryNoBackground)
        imageView?.isVisible = customization.customFlags.shouldShowLargeImages
        progress?.customize(customization)
        codeProgress?.customize(customization)
    }

    private fun setUpBulletList() {
        val title = getString(R.string.identhub_contract_signing_notice_title)
        if (title.isNotBlank()) {
            bulletList?.updateItems(
                title = title,
                items = listOf(
                    getString(R.string.identhub_contract_signing_notice1),
                    getString(R.string.identhub_contract_signing_notice2)
                ),
                titleStyle = BulletListLayout.TitleStyle.Notice,
                customization = customization
            )
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initViews()
        observeIdentificationResult()
        observableInputs()
        observeCountDownTimeEvent()
    }

    private fun initViews() {
        initFocusListener()
        initStateOfDigitInputField()
        observePhoneNumberResult()
        sendNewCode!!.setOnClickListener {
            sendNewCodeClickListener()
        }
        codeInput!!.addTextChangedListener(codeInputValidator)
        onStateOfDigitInputChanged(DEFAULT_STATE)
    }

    private fun sendNewCodeClickListener() {
        viewModel.onSendNewCodeClicked()
        codeInput!!.text.clear()
        onStateOfDigitInputChanged(DEFAULT_STATE)
        codeInput!!.addTextChangedListener(codeInputValidator)
    }

    private fun initFocusListener() {
        codeInput!!.requestFocus()
    }

    private fun initStateOfDigitInputField() {
        digitsEditTexts.add(codeInput!!)
    }

    private fun observePhoneNumberResult() {
        viewModel.getPhoneNumberResultLiveData().observe(viewLifecycleOwner, Observer { onPhoneNumberResult(it) })
    }

    private fun onPhoneNumberResult(result: Result<MobileNumberDto>) {
        if (result.succeeded && result.data != null) {
            var number = result.data!!.number
            if (number != null) {
                number = number.replace("\\d(?=\\d{4})".toRegex(), "*")
            }
            description!!.text = String.format(requireContext().resources.getString(R.string.identhub_contract_signing_description), number)
            description!!.visibility = View.VISIBLE
            progress!!.visibility = View.INVISIBLE
        }
    }

    private fun observeIdentificationResult() {
        viewModel.getIdentificationStateLiveData().observe(viewLifecycleOwner) {
            onIdentificationResultChanged(it)
        }
    }

    private fun onIdentificationResultChanged(state: ContractSigningState) {
        //todo move to use case and replace with states
        Timber.d("onIdentificationResultChanged, result: ${state}")
        when (state) {
            is ContractSigningState.SUCCESSFUL -> {
                transactionDescription!!.text = String.format(
                    getString(R.string.identhub_contract_signing_preview_transaction_info),
                    state.identificationId
                )
                onStateOfDigitInputChanged(SUCCESS_STATE)
                sharedViewModel.callOnSuccessResult()
            }
            is ContractSigningState.CONFIRMED -> {
                transactionDescription!!.text = String.format(
                    getString(R.string.identhub_contract_signing_preview_transaction_info),
                    state.identificationId
                )
                onStateOfDigitInputChanged(SUCCESS_STATE)
                sharedViewModel.callOnConfirmedResult()
            }
            is ContractSigningState.FAILED,
            is ContractSigningState.GENERIC_ERROR -> {
                onStateOfDigitInputChanged(ERROR_STATE)
            }
        }
    }

    private fun observeCountDownTimeEvent() {
        viewModel.getCountDownTimeEventLiveData().observe(viewLifecycleOwner, Observer { onCountDownTimeState(it) })
    }

    private fun onCountDownTimeState(event: Event<CountDownTime>) {
        val countDownTime = event.content
        if (countDownTime != null) {
            sendNewCode!!.visibility = if (countDownTime.isFinish) View.VISIBLE else View.GONE
            newCodeCounter!!.visibility = if (!countDownTime.isFinish) View.VISIBLE else View.INVISIBLE
            newCodeCounter!!.text = String.format(getString(R.string.identhub_contract_signing_code_expires), countDownTime.format())
        }
    }

    private fun observableInputs() {
        Timber.d("observableInputs")
        disposable = RxView.clicks(submitButton!!)
                .flatMapSingle {
                    onStateOfDigitInputChanged(LOADING_STATE)
                    Observable.fromIterable(digitsEditTexts)
                            .map { editText: EditText -> editText.text.toString() }
                            .toList()
                }
                .map { it.joinToString("") }
                .filter { token: String -> token.length == PhoneVerificationViewModel.MIN_CODE_LENGTH }
                .subscribe(
                        { token: String -> viewModel.onSubmitButtonClicked(token) },
                        { Timber.e(it, "Something went wrong") }
                )
    }

    private fun onStateOfDigitInputChanged(state: Int) {
        when(state) {
            LOADING_STATE -> {
                hideKeyboard()
                submitButton?.text = getString(R.string.identhub_verification_phone_status_verifying)
                submitButton?.isEnabled = false
                errorMessage?.visibility = View.GONE
                codeProgress?.visibility = View.VISIBLE
            }
            ERROR_STATE -> {
                submitButton?.text = getString(R.string.identhub_contract_signing_sign_action)
                submitButton?.isEnabled = isSubmitButtonEnabled()
                errorMessage?.visibility = View.VISIBLE
                codeProgress?.visibility = View.GONE
            }
            SUCCESS_STATE -> {
                submitButton?.text = getString(R.string.identhub_contract_signing_sign_action)
                submitButton?.isEnabled = false
                errorMessage?.visibility = View.GONE
                codeProgress?.visibility = View.GONE
            }
            else -> {
                submitButton?.text = getString(R.string.identhub_contract_signing_sign_action)
                submitButton?.isEnabled = isSubmitButtonEnabled()
                errorMessage?.visibility = View.GONE
                codeProgress?.visibility = View.GONE
            }
        }
        listLevelDrawables.forEach { it.level = state }
    }

    private val codeInputValidator = object : TextWatcher {
        override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {

        }

        override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
            submitButton!!.isEnabled = isSubmitButtonEnabled()
        }

        override fun afterTextChanged(s: Editable?) {
       }

    }

    private fun isSubmitButtonEnabled(): Boolean {
        return !codeInput!!.text.isNullOrEmpty() && codeInput!!.text.length == CODE_LENGTH

    }
    override fun onDestroyView() {
        disposable.dispose()
        digitsEditTexts.clear()
        currentFocusedEditText = null
        codeInput = null
        description = null
        sendNewCode = null
        newCodeCounter = null
        transactionDescription = null
        errorMessage = null
        submitButton = null
        progress = null
        codeProgress = null
        imageView = null
        bulletList = null
        super.onDestroyView()
    }

    companion object {

        const val CODE_LENGTH = 6

        const val DEFAULT_STATE = 0
        const val SUCCESS_STATE = 1
        const val ERROR_STATE = 2
        const val LOADING_STATE = 3
    }
}